source/Nevermore/Util/DataModificationQueryBuilder.cs

```diff
@@ -456,13 +464,80 @@ CommandParameterValues GetDocumentParameters(Func<DocumentMap, object> allocateI
+                sb.AppendLine(string.Join(", ", valueBlocks));
+                sb.AppendLine();
+
+                sb.AppendLine($"DELETE FROM [{data.SchemaName}].[{data.TableName}] WHERE [{data.IdColumnName}] = @{IdVariableName}");
+                sb.AppendLine($"    AND [{data.RelatedDocumentIdColumnName}] not in (SELECT Reference FROM @references)");
+                sb.AppendLine();
+
+                sb.AppendLine($"INSERT INTO [{data.SchemaName}].[{data.TableName}] ([{data.IdColumnName}], [{data.IdTableColumnName}], [{data.RelatedDocumentIdColumnName}], [{data.RelatedDocumentTableColumnName}])");
+                sb.AppendLine($"SELECT @{IdVariableName}, '{mapping.TableName}', Reference, ReferenceTable FROM @references t");
+                sb.AppendLine($"WHERE NOT EXISTS (SELECT null FROM [{data.SchemaName}].[{data.TableName}] r WHERE r.[{data.IdColumnName}] = @{IdVariableName} AND r.[{data.RelatedDocumentIdColumnName}] = t.Reference )");
+
+                for (var x = 0; x < data.Related.Length; x++)
+                    parameters.Add(relatedVariablePrefix + x, data.Related[x].relatedDocumentId);
+            }
+
+            void UseTableValuedParameter(RelatedDocumentTableData data)
+            {
+                var tableVariableName = $"{index}__relatedDocumentTableValuedParameter";
+                parameters.AddTable(tableVariableName, CreateRelatedDocumentTableValuedParameter(mapping, data));
+
+                sb.AppendLine($"DELETE FROM [{data.SchemaName}].[{data.TableName}] WHERE [{data.IdColumnName}] = @{IdVariableName}");
+                sb.AppendLine($"    AND [{data.RelatedDocumentIdColumnName}] not in (SELECT [{data.RelatedDocumentIdColumnName}] FROM @{tableVariableName})");
+                sb.AppendLine();
+
+                sb.AppendLine($"INSERT INTO [{data.SchemaName}].[{data.TableName}] ([{data.IdColumnName}], [{data.IdTableColumnName}], [{data.RelatedDocumentIdColumnName}], [{data.RelatedDocumentTableColumnName}])");
+                sb.AppendLine($"SELECT [{data.IdColumnName}], [{data.IdTableColumnName}], [{data.RelatedDocumentIdColumnName}], [{data.RelatedDocumentTableColumnName}] FROM @{tableVariableName} t");
+                sb.AppendLine($"WHERE NOT EXISTS (SELECT null FROM [{data.SchemaName}].[{data.TableName}] r WHERE r.[{data.IdColumnName}] = @{IdVariableName} AND r.[{data.RelatedDocumentIdColumnName}] = t.[{data.RelatedDocumentIdColumnName}] )");
+            }
+        }
+
+        bool ShouldUseTableValuedParameter(RelatedDocumentTableData data)
+        {
+            //80 is roughly when using a table valued parameter is faster
+            return configuration.SupportLargeNumberOfRelatedDocuments && data.Related.Length >= 80;
+        }
+
+        static TableValuedParameter CreateRelatedDocumentTableValuedParameter(DocumentMap mapping,
+            RelatedDocumentTableData data)
+        {
+            var idMetaData = new SqlMetaData(data.IdColumnName, SqlDbType.NVarChar, 400);
```

Comment:
400 seems to be used quite a few places, is that the same 400? If so, should we encapsulate it in a const??